src/pipelines/worldpop/worldpop_population.py

```diff
@@ -0,0 +1,101 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import zipfile
+from io import BytesIO
+from functools import partial
+from typing import Any, Dict, List, Tuple
+
+import requests
+from tqdm import tqdm
+from pandas import DataFrame, Series, read_csv, isnull
+
+from lib.concurrent import thread_map
+from lib.net import download
+from lib.pipeline import DataSource
+from lib.cast import age_group
+
+
+class WorldPopPopulationDataSource(DataSource):
+    """ Retrieves the requested properties from Wikidata for all items in metadata.csv """
+
+    url_strat_tpl = (
+        "ftp://ftp.worldpop.org.uk/GIS"
+        "/AgeSex_structures/Global_2000_2020/2020/"
+        "{code}/{code_lower}_{sex}_{age_bucket}_2020.tif"
+    )
+
+    url_total_tpl = (
+        "ftp://ftp.worldpop.org.uk/GIS"
+        "/Population/Global_2000_2020/2020/{code}/{code_lower}_ppp_2020.tif"
+    )
+
+    @staticmethod
+    def aggregate_population(data: Series) -> Dict[str, int]:
+        age_bucket_size = 10
+        age_bucket_count = 10
+        age_bucket_pairs = [
+            (i * age_bucket_size, (i + 1) * age_bucket_size - 1) for i in range(age_bucket_count)
+        ]
+
+        agg = {
+            "key": data["key"],
+            "population": 0,
+            "population_male": 0,
+            "population_female": 0,
+            **{f"population_age_{lo:02d}_{hi:02d}": 0 for lo, hi in age_bucket_pairs},
+        }
+        for col, val in data.iteritems():
+            # Skip over the key item
+            if col == "key":
+                continue
+
+            # Total population is the sum of all populations
+            agg["population"] += val
+
+            # Get age/sex info from column name
+            sex, age = col.split("_", 2)
+
+            # Add male and female populations separately
+            if sex == "m":
+                agg["population_male"] += val
+            if sex == "f":
+                agg["population_female"] += val
+
+            # Go over all the age buckets and add them separately
+            age_bucket = age_group(int(age)).replace("-", "_")
+            agg[f"population_age_{age_bucket}"] = val
+
+        return agg
+
+    def parse(self, sources: List[str], aux: Dict[str, DataFrame], **parse_opts) -> DataFrame:
```

Comment:
Since this is a special type of DataFrame that conforms to a specific schema, it might be worth giving it a distinct type so it's clear elsewhere in the program what type of DataFrame it is. I assume the same would apply to other pipelines.

Reply:
Discussed offline. We do this in the `DataPipeline` object via the `schema` property.